File: android/app/src/main/java/com/ooplab/exercises_fitfuel/OneEuroFilter.kt
package com.ooplab.exercises_fitfuel

import kotlin.math.PI
import kotlin.math.abs

/**
 * One Euro Filter for smoothing a single float signal in real-time.
 *
 * minCutoff: lower → smoother when still, more lag (try 1.0–2.0)
 * beta:      higher → less lag during fast movement (try 0.005–0.05)
 * dCutoff:   cutoff for the derivative low-pass (1.0 is fine)
 */
class OneEuroFilter(
    private val minCutoff: Float = 1.5f,
    private val beta: Float = 0.01f,
    private val dCutoff: Float = 1.0f
) {
    private var xHat: Float? = null
    private var dxHat = 0f
    private var lastTimestamp = 0.0

    fun filter(x: Float, timestampSec: Double): Float {
        val prev = xHat ?: run {
            xHat = x
            lastTimestamp = timestampSec
            return x
        }
        val dt = (timestampSec - lastTimestamp).toFloat().coerceAtLeast(1e-6f)
        lastTimestamp = timestampSec

        val dx = (x - prev) / dt
        dxHat = lowPass(dx, dxHat, alpha(dCutoff, dt))

        val cutoff = minCutoff + beta * abs(dxHat)
        val result = lowPass(x, prev, alpha(cutoff, dt))
        xHat = result
        return result
    }

    fun reset() {
        xHat = null
        dxHat = 0f
        lastTimestamp = 0.0
    }

    private fun alpha(cutoff: Float, dt: Float): Float {
        val r = 2f * PI.toFloat() * cutoff * dt
        return r / (r + 1f)
    }

    private fun lowPass(x: Float, prev: Float, a: Float) = a * x + (1f - a) * prev
}

File: ios/Runner/PostureEngine/OneEuroFilter.swift
import Foundation

/// One Euro Filter for smoothing a single float signal in real-time.
///
/// minCutoff: lower → smoother when still, more lag (try 1.0–2.0)
/// beta:      higher → less lag during fast movement (try 0.005–0.05)
/// dCutoff:   cutoff for the derivative low-pass (1.0 is fine)
///
/// Direct port of the Kotlin `OneEuroFilter`.
final class OneEuroFilter {
    private let minCutoff: Float
    private let beta: Float
    private let dCutoff: Float

    private var xHat: Float?
    private var dxHat: Float = 0
    private var lastTimestamp: Double = 0

    init(minCutoff: Float = 1.5, beta: Float = 0.01, dCutoff: Float = 1.0) {
        self.minCutoff = minCutoff
        self.beta = beta
        self.dCutoff = dCutoff
    }

    func filter(_ x: Float, timestampSec: Double) -> Float {
        guard let prev = xHat else {
            xHat = x
            lastTimestamp = timestampSec
            return x
        }
        let dt = max(Float(timestampSec - lastTimestamp), 1e-6)
        lastTimestamp = timestampSec

        let dx = (x - prev) / dt
        dxHat = lowPass(dx, dxHat, alpha(dCutoff, dt))

        let cutoff = minCutoff + beta * abs(dxHat)
        let result = lowPass(x, prev, alpha(cutoff, dt))
        xHat = result
        return result
    }

    func reset() {
        xHat = nil
        dxHat = 0
        lastTimestamp = 0
    }

    private func alpha(_ cutoff: Float, _ dt: Float) -> Float {
        let r = 2 * Float.pi * cutoff * dt
        return r / (r + 1)
    }

    private func lowPass(_ x: Float, _ prev: Float, _ a: Float) -> Float {
        a * x + (1 - a) * prev
    }
}
